### PR TITLE
{authenticate/proxy}: add user impersonation, manual refresh, and dashboard

### LINKED ISSUE #73: authenticate: allow manual refresh
**Is your feature request related to a problem? Please describe.**

A user tries to sign in into an application, but they are not authorized to access it because not in the only authorized group.

The admin adds them to the group for the user to be able to access the application.

The user tries to access the application but access is still refused.

**Describe the solution you'd like**

Authorization of groups (since the others are not mutable) should be refreshed after some time, if possible configurable.

**Describe alternatives you've considered**

Sign out from another application the user has access to in order to sign in again and refresh groups.

**Explain any additional use-cases**

Changing access right of a user by removing them from a group.

### LINKED ISSUE #100: proxy: provide logout link on 403 page
I use Google OAuth to login to Pomerium. I usually have more than 2 Google accounts login at the same time, but not all of them are whitelisted to access pages protected by Pomerium.

Sometimes, I would choose the wrong email address and get a 403 page. But there's no logout link on the 403 page and I have to either remember the secret URL to logout (`/.pomerium/somethingsomething`) or delete the cookie.

It would be great if you can add some description on the 403 page. E.g.: You are logged in as XXX@gmail.com. Do not intend to use this email? Click Here to login again.

Alternative without this feature: delete the cookie, or use the `/.pomerium/` link to logout (which I can barely remember 😞.

### LINKED ISSUE #110: authenticate : add impersonate user / login as support
**Is your feature request related to a problem? Please describe.**

It would be very useful, in certain circumstances, for an administrative user to impersonate or login as another user.  Some example scenarios.  

**Describe the solution you'd like**

User impersonation should allow you to temporarily sign in as a different user through pomerium. Users with impersonation permissions can impersonate all other users in their network and take any action, regardless of the impersonating user's own permission level. 

- UI flow TBD.
- Certain users or groups to impersonate other users. 
- Impersonators must appear as themselves in the change history.

**Explain any additional use-cases**

- To help another user troubleshoot an issue. If your user roles are configured differently, it's possible that your UI will look different from theirs and you'll need to impersonate the other user to be able to see what they see.
- You want to make changes on behalf of another user (for example, the other user is away on vacation and you want to manage their orders or run a report).
- You're an administrator who's setting up user roles, and you want to preview what other users will be able to see depending on the permissions you grant them.

**Prior art**

- https://help.salesforce.com/articleView?id=logging_in_as_another_user.htm&type=5
- https://confluence.atlassian.com/cloud/log-in-as-another-user-776994835.html
- https://support.google.com/admanager/answer/1241070?hl=en

### PR DESCRIPTION
**Pull request by desimone**
This PR  became bigger than I intended and includes several features and fixes beyond just impersonate user.
- Add user dashboard containing information about the current user's session. [GH-123]
- Add functionality allowing users to initiate manual refresh of their session. This is helpful when a user's access control details are updated but their session hasn't updated yet. To prevent abuse, manual refresh is gated by a cooldown (`REFRESH_COOLDOWN`) which defaults to five minutes. [GH-73]
- Add Administrator (super user) account support (`ADMINISTRATORS`). [GH-110]
- Add feature that allows Administrators to impersonate / sign-in as another user from the user dashboard. [GH-110]

Closes #73 
Closes #100
Closes #110 

**Checklist**:
- [x] updated docs
- [x] unit tests added
- [x] related issues referenced
- [x] ready for review

## Admin session
Screen Shot 2019-05-19 at 10 30 31 PM

## Non-admin user 

Screen Shot 2019-05-16 at 8 57 55 PM

## Error screen

Screen Shot 2019-05-16 at 8 58 53 PM

### CODE REVIEW

**Comment by desimone** [comment]
PR ready for review 👍 
\cc @travisgroth @nareddyt @yegle

**Comment by yegle** [comment]
One suggestion: include the `req_id` on the permission denied page, so that admin can check the log and understand why the request was rejected.

Even better if you can include a base64 encoded and encrypted protobuf message with full error reason on the permission rejected page. That way admin can just take the string and don't need to lookup in the logs.

**Comment by desimone** [comment]
@yegle great suggestion; in the interest of keeping this PR (relatively) contained, I'd like to break that into a new issue/PR. Thoughts?

**Comment by yegle** [comment]
SGTM, thanks for all the improvements :-)

File: authenticate/handlers_test.go [hunk 1]
```diff
@@ -351,58 +261,58 @@ func TestAuthenticate_SignOut(t *testing.T) {
...
-					AccessToken:  "AccessToken",
-					RefreshToken: "RefreshToken",
-					Email:        "blah@blah.com",
-
-					RefreshDeadline: time.Now().Add(10 * time.Second),
-				},
-			},
-			http.StatusBadRequest,
-			"Error"},
+		// {"good get",
```
Comment (travisgroth) [inline authenticate/handlers_test.go]: Should get rid of this block if it is deprecated

File: internal/config/options.go [hunk 2]
```diff
@@ -94,6 +94,10 @@ type Options struct {
 
 	Policies []policy.Policy
 
+	// Administrators contains a set of emails with users who have super user
+	// (sudo) access including the ability to impersonate other users' access
+	Administrators []string `envconfig:"ADMINISTRATORS"`
```
Comment (travisgroth) [inline internal/config/options.go]: Options fields need the 'mapstructure' tag to work with viper and should be LC'd going forward.

File: internal/config/options.go [hunk 3]
```diff
@@ -116,12 +120,15 @@ type Options struct {
 	// Headers to set on all proxied requests. Add a 'disable' key map to turn off.
 	Headers map[string]string `mapstructure:"headers"`
 
+	// RefreshCooldown limits the rate a user can refresh her session
+	RefreshCooldown time.Duration `mapstructure:"refresh_cooldown"`
+
 	// Sub-routes
-	Routes                 map[string]string `mapstructure:"routes"`
-	DefaultUpstreamTimeout time.Duration     `mapstructure:"default_upstream_timeout"`
+	Routes                 map[string]string `envconfig:"ROUTES"`
```
Comment (travisgroth) [inline internal/config/options.go]: mapstructure and lower case

**Comment by travisgroth** [comment]
Some minor changes needed.  Great feature, though.

File: internal/config/options.go (hunk 3 above)
Comment (desimone) [reply internal/config/options.go]: Sorry, I must have mangled that merge.

**Comment by desimone** [comment]
🤞

**Pull request merged**